### DISCUSSION TITLE
why are all platform binaries installed within `node_modules`?

### DISCUSSION THREAD

**Discussion by bmuenzenmeyer** [vercel/next.js#59821] [Help]
### Summary

I've created a new next app, and was surprised to observe the local environment's `node_modules` to be over 1 GB.  I am glad that optimized prod output with standalone mode is quite trim, but CI and developer machines suffer here. \

My team creates corporate guidance around Next.js use, and this a heavy installation cost for users on a capped internet connection, remote users with low internet speeds, or users with very tight disk space.  Does anyone know if this is intentional behavior, something that has changed, something that can be optimized, etc?

## Re-creation

```
npx create-next-app my-app
cd my-app
du -sh node_modules
```

## Info

```
npx next info

Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0: Thu Nov  2 07:43:25 PDT 2023; root:xnu-8796.141.3.701.17~6/RELEASE_ARM64_T6020
Binaries:
  Node: 18.18.0
  npm: 9.8.1
  Yarn: 1.22.21
  pnpm: 8.9.2
Relevant Packages:
  next: 14.0.4
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

## Size

```
du -sh node_modules
1.2G    node_modules
➜  my-next-app git:(main) ✗ du -sh node_modules/@next
1.1G    node_modules/@next
➜  my-next-app git:(main) ✗ du -sh node_modules/@next/*
 24K    node_modules/@next/env
132K    node_modules/@next/eslint-plugin-next
112M    node_modules/@next/swc-darwin-arm64
112M    node_modules/@next/swc-darwin-x64
113M    node_modules/@next/swc-linux-arm64-gnu
145M    node_modules/@next/swc-linux-arm64-musl
128M    node_modules/@next/swc-linux-x64-gnu
145M    node_modules/@next/swc-linux-x64-musl
 97M    node_modules/@next/swc-win32-arm64-msvc
 96M    node_modules/@next/swc-win32-ia32-msvc
129M    node_modules/@next/swc-win32-x64-msvc
```

## Research

At one point, @timneutkens stated only the supported version should be installed
https://github.com/vercel/next.js/discussions/29653#discussioncomment-1435830

Why are all of these platform-specific dependencies installed?
They seem to be direct dependencies now.

```
npm view next dependencies
{
  '@next/env': '14.0.4',
  '@swc/helpers': '0.5.2',
  busboy: '1.6.0',
  'caniuse-lite': '^1.0.30001406',
  'graceful-fs': '^4.2.11',
  postcss: '8.4.31',
  'styled-jsx': '5.1.1',
  watchpack: '2.4.0',
  '@next/swc-darwin-arm64': '14.0.4',
  '@next/swc-darwin-x64': '14.0.4',
  '@next/swc-linux-arm64-gnu': '14.0.4',
  '@next/swc-linux-arm64-musl': '14.0.4',
  '@next/swc-linux-x64-gnu': '14.0.4',
  '@next/swc-linux-x64-musl': '14.0.4',
  '@next/swc-win32-arm64-msvc': '14.0.4',
  '@next/swc-win32-ia32-msvc': '14.0.4',
  '@next/swc-win32-x64-msvc': '14.0.4'
}
```

(EDIT!: It was pointed out to me that `npm view` will display the optional dependencies here too - and you can confirm the lockfile representation correctly specifies optionalDependencies - like in https://github.com/vercel/examples/blob/main/starter/next-news/pnpm-lock.yaml#L2192-L2201)
(END EDIT)

@felixmosh looked into this at some point too via https://github.com/vercel/next.js/discussions/33398#discussioncomment-1989315

Moreover, the [contributing docs](https://github.com/vercel/next.js/blob/canary/contributing/core/developing-using-local-app.md#troubleshooting) explicitly mention using `optionalDependencies` too: 

```markdown
- If you see the below error while running `pnpm dev` with `next`:

\```
Failed to load SWC binary, see more info here: https://nextjs.org/docs/messages/failed-loading-swc
\```

Try to add the below section to your `package.json`, then run again

\```json
{
  "optionalDependencies": {
    "@next/swc-linux-x64-gnu": "canary",
    "@next/swc-win32-x64-msvc": "canary",
    "@next/swc-darwin-x64": "canary",
    "@next/swc-darwin-arm64": "canary"
  }
}
\```
```

Is that up to date? Did this change out from under the team at some point? The `package.json` seems to be programmatically generated, as I don't see any of these swc binaries in https://github.com/vercel/next.js/blob/canary/packages/next/package.json and their inclusion at the end of the list above hints to be they are added perhaps in memory and pnpm never has a chance to save / reorder them.

**Comment by bmuenzenmeyer** [comment]
So (and in reference to my edit above), considering next is doing this on purpose to put binaries in place for all possible archictures, we are left with:

- accept the size, leverage caching, pnpm, etc
- leverage something like [pnpm's `pnpfile.cjs hooks.readPackage`](https://pnpm.io/pnpmfile#hooksreadpackagepkg-context-pkg--promisepkg) to surgically remove the architectures I know I wont use
- create some post-install script in userland to prune the binaries that don't match our architecture (local and CI) - at which point we still incur the download cost/time, and only minimally improve the experience by freeing up diskspace
- ??

**Reply by 10hendersonm** [reply]
I took a crack at the `.pnpmfile.cjs` solution of stripping out binaries for different platform-architecture combos to prevent their download/installation. This cut `node_modules/` from `1.3G` to `302M` on `darwin-x64`.

> :warning: I haven't yet had a chance to validate if this continues to work when crossing environments (ex: local `darwin-x64` development on a mac :arrow_right: CI builds running on `linux-x64-gnu`).

```js
const os = require("node:os")

module.exports = {
  hooks: {
    readPackage: (pkg, _context) => {
      if (pkg.name === "next") {
        Object.keys(pkg.optionalDependencies)
          .filter(
            (optionalDependency) =>
              optionalDependency.startsWith("@next/swc-") &&
              !optionalDependency.startsWith(`@next/swc-${os.platform()}-${os.arch()}`),
          )
          .forEach((optionalDependency) => {
            console.log(`Omitting optional dependency "${optionalDependency}"`)
            delete pkg.optionalDependencies[optionalDependency]
            delete pkg.dependencies[optionalDependency]
          })
      }

      return pkg
    },
  },
}
```

# Validation

```sh
# generate a next app
pnpm create next-app test-reduced-platform-binaries \
    --typescript                                    \
    --tailwind false                                \
    --app                                           \
    --src-dir                                       \
    --eslint                                        \
    --import-alias "@/*"

cd test-reduced-platform-binaries

# get disk usage of node_modules
du -sh node_modules # 1.3G

# write a `.pnpmfile.cjs` to remove downstream `@next/swc-{platform}-{arch}*`
# `dependencies` + `optionalDependencies`
echo '
const os = require("node:os")

module.exports = {
  hooks: {
    readPackage: (pkg, _context) => {
      if (pkg.name === "next") {
        Object.keys(pkg.optionalDependencies)
          .filter(
            (optionalDependency) =>
              optionalDependency.startsWith("@next/swc-") &&
              !optionalDependency.startsWith(`@next/swc-${os.platform()}-${os.arch()}`),
          )
          .forEach((optionalDependency) => {
            console.log(`Omitting optional dependency "${optionalDependency}"`)
            delete pkg.optionalDependencies[optionalDependency]
            delete pkg.dependencies[optionalDependency]
          })
      }

      return pkg
    },
  },
}
' > .pnpmfile.cjs

# regenerate the dependency tree
rm -rf node_modules pnpm-lock.yaml
pnpm install

# get disk usage of node_modules again
du -sh node_modules # 302M

# still works
pnpm lint
pnpm dev
pnpm build && pnpm start
```

**Reply by 10hendersonm** [reply]
I tried this out and did get into issues with it - the dependencies for the other platforms won't be stored into the lockfile at all, so once you push into a CI provider running a linux container rather than a local darwin (macos) platform, you'll probably fail to run the build

**Reply by switz** [reply]
Yeah this is a huge pain, if anyone finds a solve that would be really nice

**Comment by MohammedNumair** [comment]
Create a .pnpmfile.cjs in your project root:

// .pnpmfile.cjs
const os = require('node:os');

module.exports = {
  hooks: {
    readPackage(pkg) {
      if (pkg.name === 'next' && pkg.optionalDependencies) {
        Object.keys(pkg.optionalDependencies)
          .filter(
            (d) =>
              d.startsWith('@next/swc-') &&
              !d.startsWith(`@next/swc-${os.platform()}-${os.arch()}`)
          )
          .forEach((d) => {
            // remove unwanted platform binaries
            delete pkg.optionalDependencies[d];
            delete pkg.dependencies && delete pkg.dependencies[d];
            console.log(`pnpmfile: omitting ${d}`);
          });
      }
      return pkg;
    },
  },
};



# regenerate install
rm -rf node_modules pnpm-lock.yaml
pnpm install


This approach reduced node_modules from ~1.3G → ~300M in practice.

**Comment by jpainam** [comment]
The problem still exists and the above solution didn't fix the problem.

**Comment by victorandree** [comment]
I've also run into this issue and would like to share some updates which may be helpful.

- With pnpm, I'm not seeing that other os/cpus are being installed. Even though my developer machine is an M1 Mac, building a Docker image with `--platform linux/amd64` installs only linux dependencies in the Docker image.
- However, I did notice that both glibc and musl dependencies are installed. This seems like a general issue (https://github.com/pnpm/pnpm/issues/9654).
- I've tried to filter out musl dependencies with a `.pnpmfile.cjs`, which worked for me. But I wanted to avoid this approach, because [renovatebot](https://github.com/renovatebot/renovate) when hosted doesn't support `.pnpmfile.cjs`. For Next.js, I've instead had success with `ignoredOptionalDependencies` to strip out `-musl` and `-linuxmusl`.

Most packages put `-musl` or `-linuxmusl` at the end of the package name, but some do `-linuxmusl-x64`, so you cannot just do a suffix match.

Results for an app initialized with `pnpm create next-app test-reduced-platform-binaries --typescript --tailwind false --app --src-dir --eslint --import-alias "@/*"`:

| Configuration | `du -sh node_modules` |
| :- | -: |
| Base | 605M |
| `ignoredOptionalDependencies` | 443M |
| `.pnpmfile.cjs` | 443M |

So I get identical results (and checking the contents of `node_modules/.pnpm`, they're identical), but the `.pnpmfile.cjs` also hides the other platforms from the lockfile... for better or worse.

`pnpm-workspace.yaml`:

```yaml
onlyBuiltDependencies:
  - sharp
  - unrs-resolver

# Ignore musl glibc dependencies
ignoredOptionalDependencies:
  - "*-musl*"
  - "*-linuxmusl*"
```

`.pnpmfile.cjs`:

```javascript
/**
 * Hooks into the pnpm install process by refusing to install `musl` dependencies.
 *
 * The background for this is that Next.js will install both `glibc` and `musl` dependencies, because it cannot only target one or the other. This is redundant, since we only build for glibc distributions (Debian). By stripping musl dependencies, we save ~200 MB in the final Docker image, as well as install time.
 *
 * This hook effectively removes any dependency that ends in `musl` or `linuxmusl`, when reading package.json. This means those dependencies will not be present in the lockfile. When building a Docker image, the file MUST be present when running `pnpm install` (by being copied to the image, or by being mounted). Otherwise, pnpm will compute a different checksum for the lockfile and fail to install.
 *
 * @see For information on pnpmfile, https://pnpm.io/pnpmfile
 * @see For information on Next.js, https://github.com/vercel/next.js/discussions/59821
 */
module.exports = {
  hooks: {
    readPackage(pkg) {
      if (!pkg.optionalDependencies) {
        return pkg;
      }

      for (key in pkg.optionalDependencies) {
        if (key.includes("-musl") || key.includes("-linuxmusl")) {
          delete pkg.optionalDependencies[key];
        }
      }

      return pkg;
    },
  },
};
```

`Dockerfile`:

```Dockerfile
FROM node:24 AS base
RUN npm install --global corepack@latest && corepack enable pnpm
WORKDIR /app

# Add `.pnpmfile.cjs` if using. Note: This changes the checksum of the lockfile, so rename or remove it from the repository if you're not using it, or you'll have a mismatch.
# COPY package.json .pnpmfile.cjs pnpm-*.yaml ./
COPY package.json pnpm-*.yaml ./
RUN pnpm install --frozen-lockfile
COPY . .
RUN pnpm build
RUN du -sh node_modules
```
